### ISSUE TITLE
consider using ULID for flowid

### ISSUE THREAD

**Issue by lmineiro** [zalando/skipper#251]
Recently learned about https://github.com/oklog/ulid which, without entropy, seems very performant and could offer lexicographically sortable flow Ids.

I also like the alphabet better.

Existing filter can be refactored to support a a new parameter defining the generator.

head to head benchmarking should be enough to add to the docs.

I suggest the fastest to become the default

**Comment by lmineiro** [comment]
Results on Mac OS X, 2 Ghz Core i7
current flowid
```
BenchmarkFlowIdLen8-8    	20000000	       109 ns/op
BenchmarkFlowIdLen10-8   	10000000	       127 ns/op
BenchmarkFlowIdLen12-8   	10000000	       163 ns/op
BenchmarkFlowIdLen14-8   	10000000	       165 ns/op
BenchmarkFlowIdLen16-8   	10000000	       175 ns/op
BenchmarkFlowIdLen32-8   	 5000000	       277 ns/op
BenchmarkFlowIdLen64-8   	 3000000	       477 ns/op
```
oklog/ulid
```
BenchmarkNew/WithCryptoEntropy-8         	 2000000	       937 ns/op	  17.06 MB/s
BenchmarkNew/WithEntropy-8               	20000000	        73.5 ns/op	 217.81 MB/s
BenchmarkNew/WithoutEntropy-8            	50000000	        35.5 ns/op	 451.29 MB/s
BenchmarkMustNew/WithCryptoEntropy-8     	 2000000	       954 ns/op	  16.77 MB/s
BenchmarkMustNew/WithEntropy-8           	20000000	        79.5 ns/op	 201.26 MB/s
BenchmarkMustNew/WithoutEntropy-8        	50000000	        40.0 ns/op	 399.71 MB/s
BenchmarkParse-8                         	50000000	        33.8 ns/op	 769.97 MB/s
BenchmarkMustParse-8                     	50000000		        37.6 ns/op	 692.25 MB/s
BenchmarkString-8                        	20000000	        73.3 ns/op	 218.29 MB/s
BenchmarkMarshal/Text-8                  	20000000	        60.6 ns/op	 264.19 MB/s
BenchmarkMarshal/TextTo-8                	50000000	        26.1 ns/op	 613.48 MB/s
BenchmarkMarshal/Binary-8                	300000000	         5.11 ns/op	3132.90 MB/s
BenchmarkMarshal/BinaryTo-8              	2000000000	         1.81 ns/op	8849.95 MB/s
BenchmarkUnmarshal/Text-8                	100000000	        21.9 ns/op	1188.65 MB/s
BenchmarkUnmarshal/Binary-8              	300000000	         4.79 ns/op	3341.10 MB/s
BenchmarkNow-8                           	100000000	        15.5 ns/op	 516.51 MB/s
BenchmarkTimestamp-8                     	2000000000	         0.36 ns/op	21949.72 MB/s
BenchmarkTime-8                          	2000000000	         1.10 ns/op	7282.62 MB/s
BenchmarkSetTime-8                       	2000000000	         1.09 ns/op	7331.97 MB/s
BenchmarkEntropy-8                       	200000000	         6.91 ns/op	1446.26 MB/s
BenchmarkSetEntropy-8                    	2000000000	         0.73 ns/op	13701.43 MB/s
BenchmarkCompare-8                       	200000000	         7.57 ns/op	4227.67 MB/s
```

**Comment by lmineiro** [comment]
First implementation, after refactoring a new interface for FlowID generation reveals that ULID generator would return the string representation with the following benchmark results:

```
BenchmarkFlowId/ULID-Generator-8         	 5000000	       334 ns/op
```

**Comment by aryszka** [comment]
which result of the current flowid compares to this?

**Comment by aryszka** [comment]
(ot: i like the code base of ulid :) )

**Comment by lmineiro** [comment]
To answer your question, I've refactored the old benchmarks to include the length of the ULID.
ULID returns a 23 char long flowId.

```
BenchmarkFlowIdBuiltInGenerator/23-8         	10000000	       225 ns/op
```

**Comment by lmineiro** [comment]
It's worth mentioning that the ULID generator from my current approach uses a sync.Pool with entropy readers, as suggested by https://github.com/oklog/ulid#usage

**Comment by lmineiro** [comment]
For the sake of confirmation, I made some additional tests.

1. Reusing the same entropy provider directly as a generator internal member (**not thread safe**)
```
BenchmarkFlowId/ULID-Generator-8    	10000000	       170 ns/op
```
2. sync.Pool Get() always returning the same entropy provider (**not thread safe**)
```
BenchmarkFlowId/ULID-Generator-8    	 5000000	       291 ns/op
```
3. It's exactly the same as 1) above, but uses a mutex (**thread safe**)
```
BenchmarkFlowId/ULID-Generator-8    	 5000000	       288 ns/op
```

**Comment by aryszka** [comment]
the difference between 3 and 4 got me by surprise. +1 for 4

**Comment by lmineiro** [comment]
I've been following https://github.com/golang/go/issues/14939

**Comment by lmineiro** [comment]
Implemented in #253
